concordium-consensus/src/Concordium/Skov/MonadImplementations.hs

```diff
@@ -460,114 +475,132 @@ instance SkovTimerHandlers pv (SkovPassiveHandlers pv c m) c m where
 -- * @m@: the underlying monad. Typically, this should be an instance of 'MonadIO', 'MonadLogger',
 --   and 'TimeMonad'.
 -- * @a@: the return type.
-newtype SkovT (pv :: ProtocolVersion) h c m a = SkovT {runSkovT' :: h -> SkovContext c -> StateT (SkovState c) m a}
+newtype SkovT pv h c m a = SkovT
+    { runSkovT' ::
+        PersistentTreeStateMonad
+            (SkovState c)
+            (PersistentBlockStateMonad pv (SkovContext c) (SkovTInternal pv h c m))
+            a
+    }
     deriving
-        (Functor, Applicative, Monad, MonadState (SkovState c), MonadIO, MonadLogger, TimeMonad)
-        via (ReaderT h (ReaderT (SkovContext c) (StateT (SkovState c) m)))
-
--- GlobalStateM using config abstractions
-type SkovTGSM pv h c m = GlobalStateM pv (SkovGSContext c) (SkovContext c) (SkovGSState c) (SkovState c) (SkovT pv h c m)
-type SkovTTBM pv h c m = TreeStateBlockStateM pv (SkovGSState c) (SkovGSContext c) (SkovContext c) (SkovState c) (SkovT pv h c m)
+        ( Functor,
+          Applicative,
+          Monad,
+          MonadState (SkovState c),
+          MonadIO,
+          MonadLogger,
+          TimeMonad,
+          BlockStateTypes
+        )
```

Comment:
I did this in the first iteration, but started removing `deriving via` clauses then. I got the impression we wanted to avoid them.
Personally, I did find them confusing, but less now after having dealt with them for a bit.

Reply:
I also dislike the deriving via approach. Not in general, but in this case. Since you end up deriving an instance for a type via an instance for a bigger type. Making it hard to understand what is going on and why it makes sense and screw it up by accident.